### ISSUE TITLE
Pull Migrator's `Configurable` into core

### ISSUE THREAD

**Issue by edalzell** [statamic/ideas#184]
I'd like to have a settings page for Charge, which means I need to write to the `charge.php` config file.

I have currently pulled `migrator` into my package as a dependency, but you have to a small amount of hacking (copy the `bin` folder to `public` for it to work.

Once you do that, however, it works great!

If it was pulled into core, then the migrator and the addons could use it.

Would you be open to a PR for it?

**Comment by jesseleite** [comment]
Not sure if it'll make it's way into core, but could be helpful in places.  I've thought about moving it into core to help with things like `php please multisite` (where it could automatically add merge the generated sites config in, etc) 👍

**Comment by edalzell** [comment]
Related, it is quite slow because it calls php-cs-fixer before and after every `set` call. If this is going to be pulled into core, this part should be reviewed.

**Comment by jesseleite** [comment]
How slow?  Formatting the PHP before and after is necessary for it all to work reliably.  I suspect the fact that it's calling a `php-cs-fixer` binary via `symfony/process` is why it's so slow.  Maybe there's a packagist implementation out there somewhere, but will have to look into it another time.  Works well in context of the migrator though.

**Comment by edalzell** [comment]
Yes, that's exactly why it's slow.

Yes, it's perfect for the migrator, was leaving that comment in case you choose to more that into core.

Several seconds (like 5 or so on my slow MacBook in debug) to make 13 set calls. I'm refactoring to only have 2 set calls.

**Comment by edalzell** [comment]
My refactor did not work well with the code. Back to the slow way.
<img width="922" alt="image" src="https://user-images.githubusercontent.com/6069653/82848589-03490480-9ea9-11ea-85ee-ce13e6f13414.png">

**Comment by jesseleite** [comment]
Hey @edalzell, [check out this benchmark I ran](https://gist.github.com/jesseleite/481f61978e9a32fce45e3f1cca78f152).

Though calling the `php-cs-fixer` phar via `symfony/process` is definitely the bottleneck, it doesn't seem anywhere near slow enough that you should be dealing with 20 second load times.  Curious how many times your addon is running `set` / calling php-cs-fixer?

**Comment by edalzell** [comment]
about 15 times. My machine is slow AND I have xdebug on so that likely will be a factor as well.

I can test on my new machine, which is much faster, next week when I get it set up.

**Comment by damcclean** [comment]
When I used the `Configurator` in some code recently I also had bottleneck issues like Erin said.  Here's how I implemented it in my code: https://gist.github.com/damcclean/a6d2f997493c5137036137806de7a8bf#file-statamifyimportcommand-php-L87

**Comment by jasonvarga** [comment]
You're only calling it once there but I noticed there's a `Composer::require()` in there which would definitely be slow. Maybe _that's_ the bottleneck in your case?

**Comment by edalzell** [comment]
> maybe you could optimize how you're setting configs within your addon.

I tried this as well Jesse, but when I tried to save an array I got some super weird results.

**Comment by jesseleite** [comment]
@damcclean Yeah as Jason said, it seems weird for you to feel speed issues if you're only calling it once.  Are you sure you're not feeling the `composer require` there?

@edalzell If you think xdebug might be a culprit, I'd be curious how it performs in your addon with xdebug disabled?

Maybe both of you could also use the same benchmarking technique to dump/log specific lines in your addon as well.  This example is pretty crude, but...

```php
$start = microtime(true);

// Slow line(s) of code here

Log::info((microtime(true) - $start).' seconds');
```

No rush, but it would be nice to have more concrete data though, as it seems to run fast enough in the migrator (at least here anyway).
